### ISSUE TITLE
Move transfer logic into the specific Unit subclasses

### ISSUE THREAD

**Issue by bevans2000** [mars-sim/mars-sim#474]
**Describe the bug**
The Unit Transfer logic sits inside the ```Unit``` class currently but it has logic branching to support different sub-types of Unit; but not all. For example Settlements & Building should not be transferable.

It would make the logic cleaner if we move Unit specific transfer logic into the specific Unit subclasses. For example Person would have a dedicated transfer method handling just ```Person``` transfers. The same would need to be done to ```Vehicle``` & ```Equipment```.

**Comment by bevans2000** [comment]
Not sure what is happening but after my recent pull I have 2 issues:

1. Sometimes (not often) the EVA Suit is no longer reporting in the inventory when someone is on an EVA; it was previously
2. Missions are stuck in the Disembarking phase with no leaving the vehicle

I had a couple of warnings about not being able to retrieve in the transfer method I think. Do we still enforce a retrieval before the transfer? I feel the setContainer method has to be protect to stop logic bypassing the transfer.

**Comment by bevans2000** [comment]
Also my simulation run has now stalled going into an infinite loop with a Persons walking task so the pulse is stuck just repeating the same sequence of steps.

**Comment by mokun** [comment]
I revised is settlement() and getSettlement() recently and it seems to have affected walking

**Comment by mokun** [comment]
With regsrd to the EVA suit transfer issue, it may have to do with the recent change in the embarking and disembarking methods in RoverMission

**Comment by bevans2000** [comment]
I think the problem is isInSettlement in Person., The scenario I have is a Person is in a Vehicle parked at a Settlement but outside. The person returns true for isInSettlment because the Vehicle returns true;; this means that they can now do inside Settlement tasks like tend the greenhouse which in turn prevents them from walking out of the Vehicle.
I'm thinking a rework on the transfer to clean it up and simplify it but do it on a branch.

**Comment by bevans2000** [comment]
Maybe if you are in a Vehicle even if it is parked in a Garage you can't be in the settlement until you leave the Vehicle. That would work as that is what we had before. Remind me, why was it a problem with the previous way?

**Comment by bevans2000** [comment]
Yes I can see teh changes to Person.getSettlement & Person.isInSettlement maybe causing problems. But I can't see how Persons leave the vehicle in a disembark phase.; do you know how it is triggered?

**Comment by bevans2000** [comment]
I have no idea how MissionMember are meant to leave the Vehicle during the disembarking phase. Seems that the unloading cannot start unless everyone is out of the vehicle but only no one is ever told to leave. I think the mission member should be assigned an UnloadVehcile task as part of disembarking phase. Make sense but can't understand why it isn't there already. 
I'm going to be a bit short of time over the next few days so won't be able to help much.

**Comment by mokun** [comment]
> Maybe if you are in a Vehicle even if it is parked in a Garage you can't be in the settlement until you leave the Vehicle. That would work as that is what we had before. Remind me, why was it a problem with the previous way?

My bad ! I'm now correcting `getNewLocationState()` in Vehicle to distinguish between returning the `LocationStateType` of `INSIDE_SETTLEMENT` and `WITHIN_SETTLEMENT_VICINITY`.

This way, `LocationStateType` will be reliable again and `isInSettlement()` and `getSettlement()` can depend on this enum value.

**Comment by mokun** [comment]
> how MissionMember are meant to leave the Vehicle during the disembarking phase.

Let's discuss this portion further in https://github.com/mars-sim/mars-sim/issues/488

**Comment by mokun** [comment]
It's also problematic in Person's `transfer()` in prematurely removing a person from the building. 

If person A gets inside vehicle V in a garage for loading and unloading resources, then person A is technically in 3 places at the same time, namely, being in vehicle V, in a settlement, and in a garage building.

So now, the Walk class needs to be able to process that idea.

That's why since the last few changes, the embarking and disembarking phase would have trouble processing the location of a person.
